### PR TITLE
APIGOV-16534 - add redaction library

### PR DESCRIPTION
**Pull request by jcollins-axway**
Please make sure I did not miss calling the necessary redaction where ever needed

### CODE REVIEW

**Comment by cbordier** [comment]
Maybe we should also ask Xenon team to have a look?

File: pkg/traceability/redaction/globalredaction.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+func setupSanitizeRegex(sanitizeFilters []sanitize) ([]sanitizeRegex, error) {
+	newSanitizeRegex := make([]sanitizeRegex, 0)
+	for _, filter := range sanitizeFilters {
+		if filter.KeyMatch == "" {
+			continue // ignore blank keymatches as they match nothing
+		}
+		kc, err := regexp.Compile(filter.KeyMatch)
+		if err != nil {
+
```
Comment (sbolosan) [inline pkg/traceability/redaction/globalredaction.go:46]: Your save didn't collapse this empty line?

File: pkg/traceability/redaction/globalredaction.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+		if err != nil {
+			log.Debug("Regex Error! Cannot Compile RequestHeader Remove Regex Keymatch", err)
+			return []showRegex{}, err
+		}
+
+		newShowRegex = append(newShowRegex, showRegex{
+			keyMatch: kc,
+		})
+	}
+	return newShowRegex, nil
```
Comment (sbolosan) [inline pkg/traceability/redaction/globalredaction.go:35]: Can we put a debug before we leave showing the regex value

File: pkg/traceability/redaction/globalredaction.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+			return []sanitizeRegex{}, err
+
+		}
+
+		newSanitizeRegex = append(newSanitizeRegex, sanitizeRegex{
+			keyMatch:   kc,
+			valueMatch: vc,
+		})
+	}
+	return newSanitizeRegex, nil
```
Comment (sbolosan) [inline pkg/traceability/redaction/globalredaction.go:63]: Same here... just a debug statement showing regex value

**Review by sbolosan — COMMENTED** [review]
Looks good. Some minor comments

File: pkg/traceability/redaction/globalredaction.go (hunk 2 above)
Comment (jcollins-axway) [reply pkg/traceability/redaction/globalredaction.go:35]: Yes, I am actually going to add it to the error output, meant to handle that.

**Review by dfeldick — APPROVED** [review]

**Comment by dfeldick** [comment]
Nice

**Review by dfeldick — APPROVED** [review]

File: docs/traceability/index.md [hunk 4]
```diff
@@ -751,7 +764,96 @@ func (p *EventProcessor) Process(events []publisher.Event) []publisher.Event {
 
 ```
 
+### Traceability redaction
```
Comment (jcollins-axway) [inline docs/traceability/index.md:767]: @cbordier - please review the documentation here, it will be similar to what I would put on docs.axway.com

**Review by sbolosan — APPROVED** [review]

**Review by dfeldick — APPROVED** [review]

**Review by sbolosan — APPROVED** [review]
Cool... thanks for that

**Review by dfeldick — APPROVED** [review]

File: docs/traceability/index.md (hunk 4 above)
Comment (cbordier) [reply docs/traceability/index.md:767]: @jcollins-axway this part looks strange: no formatting.
When setting up the environment variables the following is the syntax that must be used. These settings are the same as in the YAML example

TRACEABILITY_REDACTION_PATH_SHOW=[{keyMatch:"^path"}] TRACEABILITY_REDACTION_QUERYARGUMENT_SHOW=[{keyMatch:"^id$"}] TRACEABILITY_REDACTION_QUERYARGUMENT_SANITIZE=[{keyMatch:"^id$",valueMatch:"^.{0,5}"}] TRACEABILITY_REDACTION_REQUESTHEADER_SHOW=[{keyMatch:"header"}] TRACEABILITY_REDACTION_REQUESTHEADER_SANITIZE=[{keyMatch:"header",valueMatch:"."}] TRACEABILITY_REDACTION_RESPONSEHEADER_SHOW=[{keyMatch:".*"}] TRACEABILITY_REDACTION_RESPONSEHEADER_SANITIZE=[{keyMatch:"^response",valueMatch:"password"}]

Comment (jcollins-axway) [reply docs/traceability/index.md:767]: Made it a code block

Comment (jcollins-axway) [reply docs/traceability/index.md:767]: https://github.com/Axway/agent-sdk/blob/APIGOV-16534/docs/traceability/index.md#Traceability-redaction

**Review by dfeldick — APPROVED** [review]

File: docs/traceability/index.md (hunk 4 above)
Comment (jcollins-axway) [reply docs/traceability/index.md:767]: image

**Review by dfeldick — APPROVED** [review]

**Review by sbolosan — APPROVED** [review]

**Review by Atraya21 — APPROVED** [review]

**Pull request merged**